src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -0,0 +1,238 @@
+                       securityLevel:MTRStorageSecurityLevelSecure
+                         sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+
+        // Update our resumption info node list.
+        [_nodesWithResumptionInfo addObject:resumptionInfo.nodeID];
+        [_storageDelegate controller:_controller
+                          storeValue:_nodesWithResumptionInfo
+                              forKey:sResumptionNodeListKey
+                       securityLevel:MTRStorageSecurityLevelSecure
+                         sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+    });
+}
+
+- (void)clearAllResumptionInfo
+{
+    // Can we do less dispatch?  We would need to have a version of
+    // _findResumptionInfoWithKey that assumes we are already on the right queue.
+    for (NSNumber * nodeID in _nodesWithResumptionInfo) {
+        auto * oldInfo = [self findResumptionInfoByNodeID:nodeID];
+        if (oldInfo != nil) {
+            dispatch_sync(_storageDelegateQueue, ^{
+                [_storageDelegate controller:_controller
+                           removeValueForKey:ResumptionByResumptionIDKey(oldInfo.resumptionID)
+                               securityLevel:MTRStorageSecurityLevelSecure
+                                 sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+                [_storageDelegate controller:_controller
+                           removeValueForKey:ResumptionByNodeIDKey(oldInfo.nodeID)
+                               securityLevel:MTRStorageSecurityLevelSecure
+                                 sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+            });
+        }
+    }
+
+    [_nodesWithResumptionInfo removeAllObjects];
+}
+
+- (CHIP_ERROR)storeLastUsedNOC:(MTRCertificateTLVBytes)noc
+{
+    __block BOOL ok;
+    dispatch_sync(_storageDelegateQueue, ^{
+        ok = [_storageDelegate controller:_controller
+                               storeValue:noc
+                                   forKey:sLastUsedNOCKey
+                            securityLevel:MTRStorageSecurityLevelSecure
+                              sharingType:MTRStorageSharingTypeSameIdentityRequired];
+    });
+    return ok ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+}
+
+- (MTRCertificateTLVBytes _Nullable)fetchLastUsedNOC
+{
+    __block id data;
+    dispatch_sync(_storageDelegateQueue, ^{
+        data = [_storageDelegate controller:_controller
+                                valueForKey:sLastUsedNOCKey
+                              securityLevel:MTRStorageSecurityLevelSecure
+                                sharingType:MTRStorageSharingTypeSameIdentityRequired];
+    });
+
+    if (data == nil) {
+        return nil;
+    }
+
+    if (![data isKindOfClass:[NSData class]]) {
+        return nil;
+    }
+
+    return data;
+}
+
+- (nullable MTRCASESessionResumptionInfo *)_findResumptionInfoWithKey:(NSString *)key
+{
+    __block id resumptionInfo;
```

Comment:
nil check key so we don't hand nil back to the delegate?

Reply:
So this is to guard against `[NSString stringWithFormat]` returning nil?  Can do.